_posts/2020-07-20-fnn-lstm/timeseries-prediction-with-fnn-lstm.Rmd

```diff
@@ -0,0 +1,829 @@
+y_test <- test[ , (n_timesteps + 1):(2*n_timesteps), , drop = FALSE]
+
+# create tfdatasets
+ds_train <- tensor_slices_dataset(list(x_train, y_train)) %>%
+  dataset_shuffle(nrow(x_train)) %>%
+  dataset_batch(batch_size)
+
+ds_test <- tensor_slices_dataset(list(x_test, y_test)) %>%
+  dataset_batch(nrow(x_test))
+
+```
+
+Now we're ready to look at how forecasting goes on our four datasets.
+
+## Experiments
+
+### Geyser dataset
+
+People working with time series may have heard of [Old Faithful](https://en.wikipedia.org/wiki/Old_Faithful), a geyser in
+Wyoming, US that has continually been erupting every 44 minutes to two hours since the year 2004. For the subset of data
+Gilpin extracted[^3],
+
+[^3]: see dataset descriptions in the [repository\'s README](https://github.com/williamgilpin/fnn)
+
+> `geyser_train_test.pkl` corresponds to detrended temperature readings from the main runoff pool of the Old Faithful geyser
+> in Yellowstone National Park, downloaded from the [GeyserTimes database](https://geysertimes.org/). Temperature measurements
+> start on April 13, 2015 and occur in one-minute increments.
+
+Like we said above, `geyser.csv` is a subset of these measurements, comprising the first 10000 data points. To choose an
+adequate timestep for the LSTMs, we inspect the series at various resolutions:
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Geyer dataset. Top: First 1000 observations. Bottom: Zooming in on the first 200."}
+knitr::include_graphics("images/geyser_ts.png")
+```
+
+It seems like the behavior is periodic with a period of about 40-50; a timestep of 60 thus seemed like a good try.
+
+Having trained both FNN-LSTM and the vanilla LSTM for 200 epochs, we first inspect the variances of the latent variables on
+the test set. The value of `fnn_multiplier` corresponding to this run was `0.7`.
+
+```{r, eval = FALSE}
+test_batch <- as_iterator(ds_test) %>% iter_next()
+encoded <- encoder(test_batch[[1]]) %>%
+  as.array() %>%
+  as_tibble()
+
+encoded %>% summarise_all(var)
+
+```
+
+```{}
+   V1     V2        V3          V4       V5       V6       V7       V8       V9      V10
+0.258 0.0262 0.0000627 0.000000600 0.000533 0.000362 0.000238 0.000121 0.000518 0.000365
+```
+
+There is a drop in importance between the first two variables and the rest; however, unlike in the Lorenz system, `V1` and
+`V2` variances also differ by an order of magnitude.
+
+Now, it's interesting to compare prediction errors for both models. We are going to make an observation that will carry
+through to all three datasets to come.
+
+Keeping up the suspense for a while, here is the code used to compute per-timestep prediction errors from both models. The
+same code will be used for all other datasets.
+
+```{r, eval = FALSE}
+calc_mse <- function(df, y_true, y_pred) {
+  (sum((df[[y_true]] - df[[y_pred]])^2))/nrow(df)
+}
+
+get_mse <- function(test_batch, prediction) {
+  comp_df <- data.frame(test_batch[[2]][, , 1] %>%
```

Comment:
This function is hard to parse, I think splitting the pipe blocks would make it easier.

Reply:
Right! Think I found a way to make the structure more clear now :-)